src/telemetry_layer.rs

```diff
@@ -0,0 +1,106 @@
+use std::{env::var, path::PathBuf};
+use tracing::{
+    span::{self, Id, Record},
+    Event,
+};
+use tracing_appender::non_blocking::{NonBlocking, WorkerGuard};
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::{
+    fmt::{format::DefaultFields, Layer},
+    Layer as LayerTrait, Registry,
+};
+
+use crate::{
+    errors::TelemetryError, telemetry_config, telemetry_formatter::TelemetryFormatter, Result,
+};
+
+/// A `tracing` `Layer` to generate telemetry to be later consumed into InfluxDB
+///
+/// This `tracing` `Layer` formats telemetry and stores the output in a known
+/// location ($FUEL_HOME/tmp/<crate>.telemetry), ready for ingestion by an
+/// InfluxDB collector.
+pub struct TelemetryLayer {
+    pub __inner: Layer<Registry, DefaultFields, TelemetryFormatter, NonBlocking>,
+}
+
+impl TelemetryLayer {
+    /// Although public, `__new()` is only intended to be called via the
+    /// following constructor macros:
+    ///
+    /// - `fuel_telemetry::new!()`
+    /// - `fuel_telemetry::new_with_watchers!()`
+    /// - `fuel_telemetry::new_with_watchers_and_init!()`
+    pub fn __new() -> Result<(Self, WorkerGuard)> {
```

Comment:
If meant to be called internally only from crate;
```suggestion
    pub(crate) fn __new() -> Result<(Self, WorkerGuard)> {
```

Reply:
Unfortunately `__new()` is special in that it needs to be accessible from anywhere but it's mostly going to be called via `fuel_telemetry_fuel::new!()` macro. Moving it to `(crate)` will hide it from the called macro.